registry/provider.go

```diff
@@ -0,0 +1,173 @@
+
+func (dcp DefaultDoguConfigRegistryProvider) GetConfig(ctx context.Context, doguName string) (ConfigurationRegistry, error) {
+	return dcp(ctx, doguName)
+}
+
+func NewDefaultDoguConfigRegistryProvider(k8sClient ConfigMapClient) (DefaultDoguConfigRegistryProvider, error) {
+	return func(ctx context.Context, doguName string) (ConfigurationRegistry, error) {
+		provider, err := NewDoguConfigRegistry(ctx, doguName, k8sClient)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create underlying config provider: %w", err)
+		}
+
+		return provider, nil
+	}, nil
+}
+
+func NewSensitiveDoguConfigRegistryProvider[T any](k8sClient SecretClient) (ConfigRegistryProvider[T], error) {
+	sanityReg := &SensitiveDoguRegistry{}
+
+	_, ok := any(sanityReg).(T)
+	if !ok {
+		return nil, fmt.Errorf("the used interface %s is not compatible with registry %T", reflect.TypeOf(new(T)).Elem().Name(), sanityReg)
+	}
+
+	return func(ctx context.Context, doguName string) (T, error) {
+		reg, err := NewSensitiveDoguRegistry(ctx, doguName, k8sClient)
+		if err != nil {
+			return *new(T), fmt.Errorf("could not create new sensitive dogu config registry: %w", err)
+		}
+
+		// As we already did a sanity check, we can force the cast here and let it panic in case of error.
+		return any(reg).(T), nil
+
+	}, nil
+}
+
+type DefaultSensitiveDoguConfigRegistryProvider DefaultDoguConfigRegistryProvider
+
+func (dcp DefaultSensitiveDoguConfigRegistryProvider) GetConfig(ctx context.Context, doguName string) (ConfigurationRegistry, error) {
+	return dcp(ctx, doguName)
+}
+
+func NewDefaultSensitiveDoguConfigRegistryProvider(k8sClient SecretClient) (DefaultSensitiveDoguConfigRegistryProvider, error) {
+	return func(ctx context.Context, doguName string) (ConfigurationRegistry, error) {
+		provider, err := NewSensitiveDoguRegistry(ctx, doguName, k8sClient)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create underlying config registry: %w", err)
+		}
+
+		return provider, nil
+	}, nil
+}
+
+type DoguConfigWatcherProvider[T any] func(ctx context.Context, doguName string) (T, error)
+
+func (dcwp DoguConfigWatcherProvider[T]) GetDoguConfigWatcher(ctx context.Context, doguName string) (T, error) {
+	return dcwp(ctx, doguName)
+}
+
+func NewDoguConfigWatcherProvider[T any](k8sClient ConfigMapClient) (DoguConfigWatcherProvider[T], error) {
+	sanityWatcher := &DoguWatcher{}
+
+	_, ok := any(sanityWatcher).(T)
+	if !ok {
+		return nil, fmt.Errorf("the used interface %s is not compatible with watcher %T", reflect.TypeOf(new(T)).Elem().Name(), sanityWatcher)
+	}
+
+	return func(ctx context.Context, doguName string) (T, error) {
+		reg, err := NewDoguConfigWatcher(ctx, doguName, k8sClient)
+		if err != nil {
+			return *new(T), fmt.Errorf("could not create new dogu config watcher: %w", err)
+		}
+
+		// As we already did a sanity check, we can force the cast here and let it panic in case of error.
+		return any(reg).(T), nil
```

Comment:
🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked
<!-- __reviewdog__:ChAzNjYyYTk4ZTVkYjNkZTkwEgpjaGVja3N0eWxl -->